File: btrader/packages/shared/src/position-sync.ts
/**
 * WS position close-sync helpers.
 * CLOSED / position_closed bypasses event batching; OPEN must not resurrect.
 */

export function posIdOf(body: unknown): string {
  const p = body as { id?: unknown; positionId?: unknown } | null;
  return String(p?.id ?? p?.positionId ?? '');
}

export function payloadClosed(obj: unknown): boolean {
  if (!obj || typeof obj !== 'object') return false;
  const p = obj as Record<string, unknown>;
  const st = String(p.status ?? '').toUpperCase();
  const book = String(p.book ?? '').toLowerCase();
  const reason = String(p.reason ?? '').toUpperCase();
  const stateVal = String(p.state ?? '').toLowerCase();
  const event = String(p.event ?? '').toLowerCase();
  if (st === 'CLOSE_PENDING') return false;
  return (
    event === 'position_closed' ||
    p.closing === true ||
    st === 'CLOSED' ||
    book === 'closed' ||
    stateVal === 'closed' ||
    reason === 'SL_HIT' ||
    reason === 'TP_HIT'
  );
}

export function statusRank(obj: unknown): number {
  if (payloadClosed(obj)) return 3;
  const st = String((obj as { status?: unknown } | null)?.status ?? '').toUpperCase();
  if (st === 'CLOSE_PENDING') return 2;
  return 1;
}

export function forceClosed(body: unknown, id: string): Record<string, unknown> {
  const prev = body && typeof body === 'object' ? (body as Record<string, unknown>) : {};
  return {
    ...prev,
    id: prev.id ?? id,
    positionId: prev.positionId ?? prev.id ?? id,
    status: 'CLOSED',
    book: 'closed',
    closing: true,
    event: 'position_closed',
  };
}

export function coalescePosition(
  prev: unknown,
  next: unknown,
  stillClosed: (id: string) => boolean,
  rememberClosed: (id: string) => void,
): unknown {
  const id = posIdOf(next) || posIdOf(prev);
  if (!id) return next;
  const nextClosed = payloadClosed(next);
  const prevClosed = payloadClosed(prev);
  if (nextClosed || stillClosed(id) || prevClosed) {
    rememberClosed(id);
    return forceClosed(nextClosed ? next : prevClosed ? prev : next, id);
  }
  if (stillClosed(id)) {
    rememberClosed(id);
    return forceClosed(next, id);
  }
  if (statusRank(prev) > statusRank(next)) return prev;
  return next;
}

export function reconcileOpenSnapshot<T extends { id: string; status?: string }>(
  snapshot: T[] | null | undefined,
  closedIds: Iterable<string>,
): T[] {
  if (!snapshot) return [];
  const closed = closedIds instanceof Set ? closedIds : new Set(closedIds);
  return snapshot.filter((p) => {
    if (!p?.id || closed.has(p.id)) return false;
    const st = String(p.status ?? 'OPEN').toUpperCase();
    return st === 'OPEN' || st === 'CLOSE_PENDING';
  });
}

export function overlayDropsChart(id: string, payload: unknown, closedIds: Iterable<string>): boolean {
  if (!id) return false;
  const closed = closedIds instanceof Set ? closedIds : new Set(closedIds);
  if (closed.has(id)) return true;
  return payloadClosed(payload);
}

export function openPositionsRedisKey(tenantId: string, accountId: string): string {
  return `bt:${tenantId}:openpos:${accountId}`;
}

File: btrader/flutter/packages/btrader_core/lib/src/state/position_sync.dart
import '../models/position.dart';

/// True when a WS payload means the ticket is actually CLOSED (not CLOSE_PENDING).
bool positionEventIsClosed(Map<String, dynamic> data) {
  final id = '${data['id'] ?? data['positionId'] ?? ''}';
  if (id.isEmpty) return false;
  final status = '${data['status'] ?? ''}'.toUpperCase();
  if (status == 'CLOSE_PENDING') return false;
  final book = '${data['book'] ?? ''}'.toLowerCase();
  final stateVal = '${data['state'] ?? ''}'.toLowerCase();
  final reason = '${data['reason'] ?? ''}'.toUpperCase();
  final event = '${data['event'] ?? ''}'.toLowerCase();
  return status == 'CLOSED' ||
      book == 'closed' ||
      stateVal == 'closed' ||
      data['closing'] == true ||
      event == 'position_closed' ||
      reason == 'SL_HIT' ||
      reason == 'TP_HIT';
}

bool staleOpenBlocked(String id, Map<String, dynamic> data, Set<String> closedIds) {
  if (id.isEmpty || !closedIds.contains(id)) return false;
  return !positionEventIsClosed(data);
}

/// Entry / SL / TP lines stay through CLOSE_PENDING; drop on actual close.
List<Position> overlayPositions(List<Position> src, Set<String> closedIds) {
  return [
    for (final p in src)
      if (!closedIds.contains(p.id) && _workingStatus(p.status)) p,
  ];
}

bool _workingStatus(String status) {
  final s = status.toUpperCase();
  return s == 'OPEN' || s == 'CLOSE_PENDING' || s.isEmpty;
}

/// Server open-position snapshot is the reconnect source of truth.
List<Position> reconcileOpenSnapshot(
  List<Position> snapshot,
  Set<String> closedIds,
) {
  return overlayPositions(snapshot, closedIds);
}

/// Merge a live WS row into the overlay list without resurrecting CLOSED ids.
List<Position> upsertLivePosition(
  List<Position> current,
  Position next,
  Set<String> closedIds,
) {
  if (closedIds.contains(next.id) || !_workingStatus(next.status)) {
    return overlayPositions(current.where((p) => p.id != next.id).toList(), closedIds);
  }
  final out = [for (final p in current) if (p.id != next.id) p, next];
  return overlayPositions(out, closedIds);
}

List<Position> removeLivePosition(List<Position> current, String id) {
  return [for (final p in current) if (p.id != id) p];
}

Position? positionFromWs(Map<String, dynamic> data) {
  final id = '${data['id'] ?? data['positionId'] ?? ''}';
  if (id.isEmpty) return null;
  try {
    return Position.fromJson({
      ...data,
      'id': id,
      'status': data['status'] ?? 'OPEN',
      'side': data['side'] ?? 'BUY',
      'volume': data['volume'] ?? 0,
      'openPrice': data['openPrice'] ?? 0,
      'profit': data['profit'] ?? 0,
      'swap': data['swap'] ?? 0,
      'symbol': data['symbol'] ?? '',
      'accountId': data['accountId'] ?? '',
    });
  } catch (_) {
    return null;
  }
}
